src/cljam/io/vcf/util/validator.clj

```diff
@@ -0,0 +1,335 @@
+      (merge-validation-results
+       (type-validator field)
+       (number-validator num-alts field)))))
+
+(defn- make-info-field-validator [ploidy info-field-def]
+  (let [field-validator (make-field-validator ploidy info-field-def)]
+    (fn [info-key variant]
+      (let [info (get variant :info)]
+        (when-let [[_ field] (find info info-key)]
+          (let [field (if (and (some? field) (not (sequential? field)))
+                        [field]
+                        field)]
+            (-> (field-validator (count (:alt variant)) field)
+                (prepend-keys [:info]))))))))
+
+(defn- prep-info-field-defs [{:keys [ploidy] :as _validator} info]
+  (into {} (map (fn [info-field-def]
+                  (let [f (make-info-field-validator ploidy info-field-def)]
+                    [(keyword (:id info-field-def))
+                     (assoc info-field-def :fn f)])))
+        info))
+
+(defn- validate-info-fields [validator variant]
+  (->> (:info variant)
+       keys
+       (map (fn [info-key]
+              (if-let [f (get-in validator [:defs :info info-key :fn])]
+                (f info-key variant)
+                (error [:info info-key]
+                       "Info key %s not defined in meta info" info-key))))
+       (apply merge-validation-results)))
+
+(defn- make-gt-field-validator [ploidy gt-field-def]
+  (let [field-validator (make-field-validator ploidy gt-field-def)]
+    (fn [sample-name gt-key variant]
+      (let [sample (get variant sample-name)]
+        (when-let [[_ field] (find sample gt-key)]
+          (let [field (if (and (some? field) (not (sequential? field)))
+                        [field]
+                        field)]
+            (-> (field-validator (count (:alt variant)) field)
+                (prepend-keys [sample-name]))))))))
+
+(defn- prep-format-defs [{:keys [ploidy] :as _validator} fmt]
+  (into {} (map (fn [gt-field-def]
+                  (let [f (make-gt-field-validator ploidy gt-field-def)]
+                    [(keyword (:id gt-field-def))
+                     (assoc gt-field-def :fn f)])))
+        fmt))
+
+(defn- validate-format [validator {fmt :FORMAT}]
+  (if (or (nil? fmt) (sequential? fmt))
+    (->> (for [[i key] (map-indexed vector fmt)]
+           (if (keyword? key)
+             (when-not (get-in validator [:defs :format key])
+               (error-on-bcf validator
+                             [:FORMAT i]
+                             "Genotype key %s not defined in meta info" key))
+             (error [:FORMAT i]
+                    (str "Every genotype key in :FORMAT must be keyword, "
+                         "but got " (pr-str key)))))
+         (apply merge-validation-results))
+    (error [:FORMAT]
+           (str ":FORMAT must be a sequence of keywords or nil, but got "
+                (pr-str fmt)))))
+
+(defn- validate-samples [validator variant]
+  ;; TODO: The presence of the GT field should be checked if :FORMAT is present
```

Comment:
[VCFv4.4 Specification](https://github.com/samtools/hts-specs/blob/9ea3a9d1fe1f18579f9a429ef224a8d05dbc48f1/VCFv4.4.tex#L443) says:
> The first key must always be the genotype (GT) if it is present.

There seems to be a discrepancy between this `TODO:` comment and the specification. Is this comment written with that in mind?

Reply:
I was not really sure what you actually meant by "discrepancy", but anyway I just added some code to check the presence of `:GT` as the first genotype field, rather than leaving it to be a TODO.

Reply:
Thanks for adding the new commit 👍  I have confirmed that.
What I'm asserting is that the VCF specification permits the absence of the GT field.
Is there a practical reason why the GT field is considered essential?

Reply:
> What I'm asserting is that the VCF specification permits the absence of the GT field.

Ah, right 😂 In theory, `:FORMAT` is allowed to be empty, so technically my comment wasn't accurate.

> Is there a practical reason why the GT field is considered essential?

I'm not the spec author, so I don't know what the real reason is, but I guess one of the reasons is that if you don't know the genotypes, you can’t tell how many values are necessary for the fields of `Number G` (which require as many values as the genotypes).